### PR TITLE
Changed message history retrieval methods to use IDs instead of message objects

### PR DESCRIPTION
**Pull request by jcryer**
# Summary
Changed DiscordChannel.LastMessageId into a DiscordMessage object and renamed to DiscordChannel.LastMessage.

# Details
Proposing a fix for situations like this (now that GetMessagesAsync has been deprecated):
```
await c.GetMessagesBeforeAsync(await c.GetMessageAsync(c.LastMessageId), 5);
```
By changing to a system like this:
```
await c.GetMessagesBeforeAsync(c.LastMessage, 5);
```

This is where `c` is a DiscordChannel, and is not the channel the command has been issued in.

# Changes proposed
* Change LastMessageId into LastMessage (DiscordMessage object)

Pretty simple change, just thought it would simplify these situations.

### CODE REVIEW

**Comment by uwx** [comment]
Very nice, but wouldn't this only work if the message were in the cache?

**Comment by jcryer** [comment]
I'm pretty sure LastMessageId was only set on MessageCreated event, and in that situation the entire DiscordMessage is sent, right?

**Comment by uwx** [comment]
Alright, that makes sense.

**Comment by Emzi0767** [comment]
You should not rename it. Instead you should internalise (preferably not though) the ID property, and then just expose a message property which retrieves from cache, if you make such a change. 

The reason is that this could easily lead to memory leaks if you permanently keep the message object in memory. And the reason hiding/replacing ID is not a change I like is because the last message might be missing from cache.

**Comment by Emzi0767** [comment]
And whose idea was it to make the message getting methods work on objects rather than IDs?

**Comment by uwx** [comment]
uhhhhhhh

**Comment by uwx** [comment]
well it's an object oriented programming language, for one

**Comment by Emzi0767** [comment]
And how exactly are you supposed to retrieve an object which you don't have? It's the same as making the single message method require a message object to retrieve the same object.

**Comment by uwx** [comment]
Isn't it reasonable to assume you get a message from any of the events that give you message objects?

**Comment by Emzi0767** [comment]
It's not, particularly for traversing historical data.

**Comment by jcryer** [comment]
Oh right, so if no message has been created for a long time in a channel, the DiscordMessage won't be in the cache...????
But then we're assuming the ID remains in the cache, aren't we?
Correct me if i'm wrong.

**Comment by Emzi0767** [comment]
ID is a number. It's assigned to the channel object; cache is irrelevant here.

**Comment by jcryer** [comment]
Oh right, my bad.

Anyway, `Discord.ApiClient.GetChannelMessagesAsync()` takes the ID anyway, there is 0 point in taking in the full object.

**Comment by jcryer** [comment]
I basically just reverted it to use ulongs, i see no point in having DiscordMessages - it just complicates any situation I can actually think of that you would actually use any of the `GetMessages` methods.

File: DSharpPlus/Entities/DiscordChannel.cs [hunk 1]
```diff
@@ -92,7 +92,7 @@ public IReadOnlyList<DiscordOverwrite> PermissionOverwrites
         public string Topic { get; internal set; } = "";
 
         /// <summary>
-        /// Gets the ID of the last message sent in this channel. This is applicable to text channels only.
+        /// Gets the id of the last message sent in this channel. This is applicable to text channels only.
```
Comment (Emzi0767) [inline DSharpPlus/Entities/DiscordChannel.cs]: Please re-capitalize

**Comment by Quahu** [comment]
Why not just have an *overload* with ulong

**Pull request merged**